Update index usage docs (#15679)
MaterializeInc/materialize: doc/user/content/ops/optimization.md

  - /ops/speed-up/
---

> Use indexes to speed up queries. Improvements can be significant, reducing some query times down to single-digit milliseconds. In particular, when the query filters only by the fields of an index, then Materialize can look up just the desired values from the index.

Reviewer: ```suggestion
## Indexes

Like in any standard relational database, you can use [indexes](/overview/key-concepts/#indexes) to optimize query performance in Materialize. Improvements can be significant, reducing some query times down to single-digit milliseconds.
```

`ORDER BY` and `LIMIT` aren't clauses that benefit from an index.

### `WHERE`
Speed up a query involving a `WHERE` clause with equality comparisons to literals, using the following table as a guide. (`collection` can be a source, view, materialized
view, or table. `$1`, `$2`, etc. mean literals, e.g., `42`, or `'foo'`.)

Clause                    | Index                                |
--------------------------|--------------------------------------|
`WHERE x = $1`            | `CREATE INDEX ON collection (x);`    |
`WHERE x IN ($1, $2, $3)` | `CREATE INDEX ON collection (x);`        |
`WHERE (x, y) IN (($1, $2), ($3, $4))` | `CREATE INDEX ON collection (x, y);`     |
`WHERE x = $1 AND y = $2` | `CREATE INDEX ON collection (x, y);`     |
`WHERE (x = $1 AND y = $2) OR (x = $3 AND y = $4`) | `CREATE INDEX ON collection (x, y);`     |
`WHERE x * 2 = $1`        | `CREATE INDEX ON collection (x * 2);`    |
`WHERE upper(x) = $1`     | `CREATE INDEX ON collection (upper(x));` |

> You can verify that the index is being used for a lookup with `EXPLAIN`, which should show `lookup`:

Reviewer 2: What about turning this excellent suggestion into an active voice?

Examples:
```suggestion
You can verify an index usage in a lookup using `EXPLAIN`. Search for `lookup` after the index name to confirm its usage. 
```

```suggestion
Use `EXPLAIN` to verify that a lookup query uses an index. Look for `ReadExistingIndex`/`lookup` to assert an index's usage:
```

Author: Thanks, I rewrote the sentences!

```
materialize=> EXPLAIN SELECT * FROM foo WHERE x = 6 AND y = 8;
                            Optimized Plan
----------------------------------------------------------------------
 Explained Query (fast path):                                        +
   Project (#0, #1)                                                  +
     ReadExistingIndex materialize.public.foo_x_y lookup value (6, 8)+
                                                                     +
 Used Indexes:                                                       +
   - materialize.public.foo_x_y                                      +
```

> **Matching multi-column indexes to multi-column `WHERE` clauses:** In general, your index key should exactly match the columns that are constrained in your `WHERE` clause. In more detail:

Reviewer 2: What if **Matching multi-column indexes to multi-column `WHERE` clauses** goes as a title and frees you to describe the section?

```suggestion
#### Matching multi-column indexes and additional details 

In general, your index key should match the columns constrained in your `WHERE` clause. In more detail ..:
```

Author: I made it into a title.

> - **Attention:** If your `WHERE` clause constrains fewer fields than your index key includes, then **the index will not be used**. For example, an index on `(x, y)` cannot be used to speed up `WHERE x = $1`.

Reviewer 2: This section is very good cause it encapsulates the key details while using queries and understanding Materialize behavior. Still, I'm unsure how to display the text structure, but we can leave it for the future.

- If you have an OR not all of whose arguments constrain the same fields, then create an index for the intersection of the constrained fields. For example, if you have `WHERE (x = $1 AND y = $2) OR x = $3`, then create an index just on `x`.
- If you have an OR whose arguments constrain completely disjoint sets of fields (e.g., `WHERE x = $1 OR y = $2`), then try to rewrite your query using a UNION, where each argument of the UNION has one of the original OR arguments (with taking into account possible duplicates).

### `JOIN`
Speed up a `JOIN` query by indexing the join keys:

Clause                                      | Index                                                                       |
--------------------------------------------|-----------------------------------------------------------------------------|
`FROM view V JOIN table T ON (V.id = T.id)` | `CREATE INDEX ON view (id);` <br /> `CREATE INDEX ON table (id);`           |

> For joins between more than two collections, you should strive for a *delta join*, which is typically possible when you index all the join keys:

Author: I added
"(a special join implementation of Materialize for making multi-way streaming joins memory-efficient)"

         ArrangeBy keys=[[#0], [#1]]             +
           Get materialize.public.t2             +
         ArrangeBy keys=[[#0]]                   +
           Get materialize.public.t3             +
                                                 +
 Used Indexes:                                   +
   - materialize.public.t1_y                     +
   - materialize.public.t3_x                     +
   - materialize.public.t2_x                     +
   - materialize.public.t2_y                     +
```
Delta joins have the advantage that they will use negligible additional memory outside the explicitly created indexes. For more details, see [Maintaining Joins using Few Resources](https://materialize.com/blog/maintaining-joins-using-few-resources).

If your query filters one or more of the join inputs by a literal equality (e.g., `t1.y = 42`), then put one of those inputs first in the `FROM` clause. This can especially speed up one-off SELECT queries (i.e., queries that are not long-running dataflows, continuously updating their results), because this will avoid scanning any of the input collections: all of them will be accessed only by index lookups.

### `GROUP BY`

> Speed up a query using a `GROUP BY` by indexing the aggregation keys:

Reviewer 2: I can't add the suggestion here :(

> Speed up a query using a `GROUP BY` by indexing the aggregation keys:

Author: Fixed.

(This might still change, based on Marta's input.)

 Used Indexes:                                                              +
   - materialize.public.foo_x_y                                             +
```

#### Matching multi-column indexes to multi-column `WHERE` clauses

In general, your index key should exactly match the columns that are constrained in your `WHERE` clause. In more detail:
- **Attention:** If your `WHERE` clause constrains fewer fields than your index key includes, then **the index will not be used**. For example, an index on `(x, y)` cannot be used to speed up `WHERE x = 7`.
- If your `WHERE` clause constrains more fields than your index key includes, then the index might still provide some speedup, but it won't necessarily be optimal: In this case, the index lookup is performed using only those constraints that are included in the index key, and the rest of the constraints will be used to subsequently filter the result of the index lookup.
- If you have an `OR` not all of whose arguments constrain the same fields, then create an index for the intersection of the constrained fields. For example, if you have `WHERE (x = 51 AND y = 'bbb') OR (x = 76 AND z = 9)`, then create an index just on `x`.

> - If you have an `OR` whose arguments constrain completely disjoint sets of fields (e.g., `WHERE x = 5 OR y = 'aaa'`), then try to rewrite your query using a `UNION` (or `UNION ALL`), where each argument of the `UNION` has one of the original `OR` arguments (with taking into account possible duplicate result rows from the intersection).

Author: The trouble is that blindly changing to either `UNION` or `UNION ALL` is wrong here:
1. If you use `UNION ALL`, then you will get extra duplicates for those records that pass both predicates (both arguments of the original OR).
2. If you use `UNION`, you lose duplicates everywhere.

It's mostly because of these complications that the optimizer is not doing this automatically at the moment (https://github.com/MaterializeInc/materialize/issues/4229). The user might be in a better better position to know whether 1. or 2. is appropriate for her use case:
1. might be good if the intersection is empty, or if duplicates are not an issue downstream.
2. might be good if there were no duplicates to begin with, or if losing duplicates is not an issue downstream.

So I think an extra warning here about duplicates is not redundant.

> - If you have an `OR` whose arguments constrain completely disjoint sets of fields (e.g., `WHERE x = 5 OR y = 'aaa'`), then try to rewrite your query using a `UNION` (or `UNION ALL`), where each argument of the `UNION` has one of the original `OR` arguments (with taking into account possible duplicate result rows from the intersection).

Reviewer: Oh, what I meant is that I'd expect users to be aware of whether they need to use one or the other depending on whether they need to care or not about duplicates. But I understand your point, it definitely doesn't hurt!

### `JOIN`
Speed up a `JOIN` query by indexing the join keys:

Clause                                      | Index                                                                       |
--------------------------------------------|-----------------------------------------------------------------------------|
`FROM view V JOIN table T ON (V.id = T.id)` | `CREATE INDEX ON view (id);` <br /> `CREATE INDEX ON table (id);`           |

> For joins between more than two collections, you should strive for a *delta join* (a special join implementation of Materialize for making multi-way streaming joins memory-efficient), which is typically possible when you index all the join keys:

Author: I changed it to "inputs". (They are not necessarily a storage collections. For example, a join input might be an index o na non-materialized view.)
